test/test_meta.py

```diff
@@ -668,129 +118,875 @@ def meta_tensor(t):
+                t.is_sparse_csr, t.is_sparse, t.is_mkldnn, t.is_quantized,
+                t.is_nested, torch._is_functional_tensor(t),
+                # these are supported in meta conversion but the fallbacks
+                # don't work
+                t.is_neg(), t.is_conj(),
+                # conjugate fallback does not support meta tensors
+                t.dtype in (torch.complex128, torch.complex64),
+            ]):
+                # TODO: sparse should support meta
+                # NB technically to('meta') does work but our logging
+                # instrumentation will see the meta conversions and the
+                # tests all break so we just exclude this.  In any case
+                # the to conversion isn't really right anyhow.
+                self.miss += 1
                 return t
+            elif any([
+                t.device.type in ("lazy", "meta"), t.is_complex(),
+                # We need a way to test if a tensor is batched but there
+                # is no official APi to do it
+                # torch._C._is_batched(t),
+            ]):
+                # TODO: this stuff should support storage
+                # (well, maybe not batched)
+                self.hit += 1
+                return t.to("meta")
             else:
-                # non-Tensor types don't count as hit or miss
-                return t
+                self.hit += 1
+                r = self.meta_tensor(t)
+                if type(t) is torch.nn.Parameter:
+                    r = torch.nn.Parameter(r, requires_grad=r.requires_grad)
+                return r
+        elif torch.overrides.is_tensor_like(t):
+            # Blindly converting tensor subclasses to meta can cause
+            # unpredictable problems; e.g., FX tests will trace meta
+            # tensors into their trace / some subclasses don't correctly
+            # support meta.  Trying to YOLO this is more trouble than it's
+            # worth.
+            self.miss += 1
+            return t
+        else:
+            # non-Tensor types don't count as hit or miss
+            return t
 
-        do_meta = (
-            (self.run_excludes_anyway or func not in meta_exclude_set) and
-            not torch.jit.is_tracing() and
-            not isinstance(func, torch.ScriptMethod)
-        )
 
-        if do_meta:
-            try:
-                meta_args = tree_map(to_meta, args)
-                meta_kwargs = tree_map(to_meta, kwargs)
-            except Exception as e:
-                raise RuntimeError(
-                    f"failed to convert args to meta; "
-                    f"originally (*{args}, **{kwargs})") from e
+def assert_ref_meta_equal(test_case, meta_rs, rs, msg_callable):
+    flat_meta_rs, _ = tree_flatten(meta_rs)
+    flat_rs, _ = tree_flatten(rs)
+    test_case.assertEqual(len(flat_meta_rs), len(flat_rs))
+    for i, meta_r, r in zip(range(len(flat_rs)), flat_meta_rs, flat_rs):
+        def test_assert(cond, msg):
```

Comment:
is it worth asserting on the shape of the tree here ? I've seen instances in past where say a meta decomposition returned a list but the corresponding function returned a tuple

Reply:
Yes I agree